Libraries/Opc.Ua.Client/Session/SessionAsync.cs

```diff
@@ -1051,6 +1061,300 @@ List<ServiceResult> errors
+
+                    passCount++;
+
+                } while (nodesToBrowseForPass.Count > 0);
+            }
+            catch (Exception ex)
+            {
+                Utils.LogError(ex, "ManagedBrowse failed");
+            }
+
+            return (result, errors);
+        }
+
+        /// <summary>
+        /// Used to pass on references to the Service results in the loop in ManagedBrowseAsync.
+        /// </summary>
+        private class ReferenceWrapper<T>
+        {
+            public T reference { get; set; }
+        }
+
+        /// <summary>
+        /// Call the browse service asynchronously and call browse next,
+        /// if applicable, immediately afterwards. Observe proper treatment
+        /// of specific service results, specifically
+        /// BadNoContinuationPoint and BadContinuationPointInvalid
+        /// </summary>
+        /// <param name="view"></param>
+        /// <param name="nodeIds"></param>
+        /// <param name="maxResultsToReturn"></param>
+        /// <param name="browseDirection"></param>
+        /// <param name="referenceTypeId"></param>
+        /// <param name="includeSubtypes"></param>
+        /// <param name="nodeClassMask"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        private async Task<(
+            List<ReferenceDescriptionCollection>,
+            IList<ServiceResult>
+            )>
+            BrowseWithBrowseNextAsync(
+            ViewDescription view,
+            List<NodeId> nodeIds,
+            uint maxResultsToReturn,
+            BrowseDirection browseDirection,
+            NodeId referenceTypeId,
+            bool includeSubtypes,
+            uint nodeClassMask,
+            CancellationToken ct = default
+            )
+        {
+            var result = new List<ReferenceDescriptionCollection>();
+
+            (
+                _,
+                ByteStringCollection continuationPoints,
+                IList<ReferenceDescriptionCollection> referenceDescriptions,
+                IList<ServiceResult> errors
+            ) =
+            await BrowseAsync(
+                null,
+                view,
+                nodeIds,
+                maxResultsToReturn,
+                browseDirection,
+                referenceTypeId,
+                includeSubtypes,
+                nodeClassMask,
+                ct).ConfigureAwait(false);
+
+
+            result.AddRange(referenceDescriptions);
+
+            // process any continuation point.
+            var previousResults = result;
+            var errorAnchors = new List<ReferenceWrapper<ServiceResult>>();
+            var previousErrors = new List<ReferenceWrapper<ServiceResult>>();
+            foreach (var error in errors)
+            {
+                previousErrors.Add(new ReferenceWrapper<ServiceResult> { reference = error });
+                errorAnchors.Add(previousErrors.Last());
+            }
+
+            byte[] nullByte = new byte[16];
+
```

Comment:
has the continuation token always this length?

Reply:
this we discussed in chat. The ReferenceServer sends a continuation point of the form byte[16] with all entries initial, when it returns the status code BadNoContinuationPoints. In some places (in the client samples and the node cache) the status code was ignored and the continuation points were retrieved from the Browse(Next)Response which were invalid and sent in a BrowseNext to thes server (which, of course, replied with BadInvalidContinuationPoint). Ths nullByte and the check for the nullByte is not needed if the status code bad is respected. 
I removed it, accordingly.